Start process instance at specific element

remcowesterhoud opened the pull request:
## Description

<!-- Please explain the changes you made here. -->
This PR will make it possible to start a PI anywhere in the process. It will look in the start instructions on which element(s) the process should be started. It will start by activating the process itself. Using recursing it will then activate all the flow scopes (e.g. subprocesses) that are required for the element in the start instruction. It will do this in the correct order (top to bottom). Lastly it will send an activate command for the element(s) in the start instructions.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #9390 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

remcowesterhoud commented:
Requesting a review from both of you as it is important to do some knowledge sharing here.

pihme commented on engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:
```diff
@@ -79,8 +89,12 @@ public boolean onCommand(
+      commandWriter.appendFollowUpCommand(
+          processInstanceKey, ProcessInstanceIntent.ACTIVATE_ELEMENT, processInstance);
+    } else {
+      activateElementsForStartInstructions(record, process, processInstanceKey, processInstance);
```
:wrench: You could pass in just the start instructions. The record is only used to get to the start instructions

pihme commented on engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:
```diff
@@ -199,4 +214,95 @@ private DeployedProcess getProcess(final long key, final CommandControl controll
+              final long elementKey = keyGenerator.nextKey();
+              final ProcessInstanceRecord elementRecord =
+                  createProcessInstanceRecord(process, processInstanceKey, elementId, flowScopeKey);
+              commandWriter.appendFollowUpCommand(
```
:x: I think this could activate an element twice. E.g. given this process:
![image](https://user-images.githubusercontent.com/14032870/172837320-d0eec1c8-15d7-4b8d-b0e5-10275564077e.png)

With  `client.createProcessInstance().startBefore("task").startBefore("subprocess").send().join()` the sub process would be activated twice. The first time as part of `activateFlowScopes()` for `task` and the second time when the second start instruction is processed.

remcowesterhoud replied:
I think you are right. It's a bit of an odd scenario. What would the expected behaviour be in this case? 
Send a command to activate the subprocess, and then send a command to activate the task. This would sort of create 2 process instances in one 😅 Then the task would be activated twice.

pihme replied:
It is valid user input though. 

Could be a misunderstanding. Maybe the customer didn't know we will activate the surrounding event scopes and so they want to do it explicitly

saig0 replied:
> With client.createProcessInstance().startBefore("task").startBefore("subprocess").send().join() the sub process would be activated twice. The first time as part of activateFlowScopes() for task and the second time when the second start instruction is processed.

Yes, it would create the subprocess twice. For me, this is the expected behavior. 

The rule for process instance modification is: it creates a new element instance in the existing scope instance. If no scope instance exists yet then a new scope instance is created.

The start-process-anywhere command follows the same rule. 

Independent of the order of the two instructions, the scope instance of the subprocess doesn't exist yet. As a result, it will create a new scope instance of the subprocess twice.

I'll create a new test case for it.

pihme commented on engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:
```diff
@@ -199,4 +214,95 @@ private DeployedProcess getProcess(final long key, final CommandControl controll
+   * process is already activated, so the activation for this is skipped (it is present in the
+   * activatedFlowScopeIds map). From here it traverses the flow scopes back down to the desired
+   * start element and activates all flow scopes in order. The desired start element is not
+   * activated here as an activate command is send for this later on.
```
```suggestion
   * activated here as an activate command is sent for this later on.
```

pihme commented on engine/src/main/java/io/camunda/zeebe/engine/state/instance/ElementInstance.java:
```diff
@@ -197,11 +197,14 @@ public long getActiveSequenceFlows() {
-      throw new IllegalStateException(
-          "Not expected to have an active sequence flow count lower then zero!");
+    if (getActiveSequenceFlows() > 0) {
+      activeSequenceFlowsProp.decrement();
```
:thinking: Do not understand this change

pihme reviewed (changes requested):
Good work, but I do have some concerns in the details. Also, I have not reviewed the test yet.

pihme commented on engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:
```diff
@@ -199,4 +214,95 @@ private DeployedProcess getProcess(final long key, final CommandControl controll
+      final DeployedProcess process,
+      final long processInstanceKey,
+      final ProcessInstanceRecord processInstance) {
+    stateWriter.appendFollowUpEvent(
```
Or maybe I am wrong. I routinely get confused which bit is done where.

remcowesterhoud replied:
I believe the variable behaviour is part of this Processor. One of the first things that happens when receiving this command is merging the variables from the command into the scope of the process instance.
```java
    final long processInstanceKey = keyGenerator.nextKey();
    if (!setVariablesFromDocument(
        controller, record, process.getKey(), processInstanceKey, process.getBpmnProcessId())) {
      return true;
    }
```
```java
  private boolean setVariablesFromDocument(
      final CommandControl<ProcessInstanceCreationRecord> controller,
      final ProcessInstanceCreationRecord record,
      final long processDefinitionKey,
      final long processInstanceKey,
      final DirectBuffer bpmnProcessId) {
    try {
      variableBehavior.mergeLocalDocument(
          processInstanceKey,
          processDefinitionKey,
          processInstanceKey,
          bpmnProcessId,
          record.getVariablesBuffer());
    } catch (final MsgpackReaderException e) {
      Loggers.PROCESS_PROCESSOR_LOGGER.error(ERROR_INVALID_VARIABLES_LOGGED_MESSAGE, e);
      controller.reject(
          RejectionType.INVALID_ARGUMENT,
          String.format(ERROR_INVALID_VARIABLES_REJECTION_MESSAGE, e.getMessage()));

      return false;
    }

    return true;
  }
```

remcowesterhoud replied:
I do think a test to verify this is warranted so I will add this.

saig0 commented on engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:
```diff
@@ -0,0 +1,423 @@
+                .limitToProcessInstanceCompleted())
+        .extracting(record -> record.getValue().getBpmnElementType(), Record::getIntent)
+        .containsSubsequence(
+            Tuple.tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            Tuple.tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            Tuple.tuple(BpmnElementType.END_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            Tuple.tuple(BpmnElementType.END_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATED));
```
:wrench: We could add a static import for `tuple()` to reduce the code :) 

```suggestion
            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATING),
            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
            tuple(BpmnElementType.END_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATING),
            tuple(BpmnElementType.END_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATED));
```

saig0 commented on engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:
```diff
@@ -0,0 +1,423 @@
+                .limit(1))
+        .hasSize(1);
+
+    Assertions.assertThat(
+            RecordingExporter.variableRecords(VariableIntent.CREATED)
+                .withProcessInstanceKey(key)
+                .getFirst())
+        .extracting(Record::getValue)
+        .extracting(
+            VariableRecordValue::getScopeKey,
+            VariableRecordValue::getName,
+            VariableRecordValue::getValue)
+        .containsExactly(key, "variable", "\"value\"");
```
:wrench: We could limit the stream to the process instance instead to verify that the variable is created only for the process instance and not more. For example:

```
RecordingExporter.
  records()
  .limitToProcessInstance(key)
  .variableRecords()
  // ...
```

remcowesterhoud replied:
A limit like this doesn't work as `limitToProcessInstance` also filters on specific `ProcessInstanceIntents`, thus not returning the variable records. I don't see another way to limit, am I missing something?

saig0 replied:
Not true. It limits the stream based on the process instance records of the given process instance key but it returns the variable records.

However, I just realized that it doesn't work because the variables are written before the process instance record of the process is written. 

But we could write it in the following way:

```
Assertions.assertThat(
            RecordingExporter.records()
                .limit(
                    // extract the condition into a method like limitToProcessInstance()
                    r ->
                        r.getKey() == key
                            && r.getIntent() == ProcessInstanceIntent.ELEMENT_COMPLETED)
                .variableRecords())
        .hasSize(1)
        .extracting(Record::getValue)
        .extracting(
            VariableRecordValue::getScopeKey,
            VariableRecordValue::getName,
            VariableRecordValue::getValue)
        .containsExactly(tuple(key, "variable", "\"value\""));
```

saig0 commented on engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:
```diff
@@ -0,0 +1,423 @@
+import org.junit.Rule;
+import org.junit.Test;
+
+public class CreateProcessInstanceAnywhereTest {
```
:wrench: We should add another test to verify the properties of the `activate` command and the `activating` + `activated` events that are written by the processor.

If the processor writes the records not correctly then it can be hard to root cause the issue. For example, a wrong flow scope key caused the process to not complete.

saig0 reviewed (approved):
@remcowesterhoud good job. I like it :+1: 

I have only a few minor suggestions. Nothing major. 

Don't worry If you don't have time to finish the PR today. I can take over :handshake:

remcowesterhoud commented on engine/src/test/java/io/camunda/zeebe/engine/util/client/ProcessInstanceClient.java:
```diff
@@ -67,6 +68,12 @@ public ProcessInstanceCreationClient withVariable(final String key, final Object
       return this;
     }
 
+    public ProcessInstanceCreationClient withStartInstruction(
+        final ProcessInstanceCreationStartInstruction instruction) {
+      processInstanceCreationRecord.addStartInstruction(instruction);
+      return this;
+    }
```
I don't agree with this. The record we are creating contains a `startInstruction` property. It makes perfect sense to me to have a `withStartInstruction` method which will add this instruction. Whether it is start before or after an element is irrelevant. Especially since there is only one possibility. If we ever decide to support multiple options they should be added to the start instruction that is passed.

I don't like to specify it in the test names either. We don't know the concept of starting before or after an element in Zeebe, so why add it to the test name? I could understand doing it in the Java client for backwards compatibility down the road, but here I feel it's unnecessary.

saig0 replied:
Okay :smile: I don't force you to change.

remcowesterhoud commented:
I have addressed most of your comments. The only remaining things are the unresolved comments. They should be fairly easy to implement. @saig0 I will leave it in your hands 😄

saig0 commented:
bors merge

saig0 commented:
@pihme please have another look. I need your approval to merge the PR :heavy_check_mark:

pihme commented on broker/src/test/java/io/camunda/zeebe/broker/engine/RecordVersionTest.java:
```diff
@@ -68,7 +68,7 @@ public void processInstanceRecordsShouldHaveBrokerVersion() {
 
     // then
-    assertThat(RecordingExporter.records().limitToProcessInstance(processInstanceKey))
+    assertThat(RecordingExporter.records().betweenProcessInstance(processInstanceKey))
```
:x: I find this naming confusing. Maybe this could be rewritten to
`betweenProcessInstanceEvents(processInstanceKey, List.of(ProcessInstanceIntent.ELEMENT_ACTIVATING), List.of(ProcessInstanceEvent.ELEMENT_COMPLETED, ProcessInstanceEvent.ELEMENT_TERMINATED))`

Although _events_ is also ambiguous. I am looking for a word that describes a point in the lifecycle of a process instance.

saig0 replied:
I accept your objections regarding the naming but I don't see a reason to block this PR. We can change the name later if we have a better idea.

pihme replied:
Then let's create an issue for the renaming, so we don't forget it. And then this PR can go ahead.

saig0 replied:
I will not create an issue for the renaming. If someone has a better idea for the name then we can just create a PR. I don't see that this name is so important.

pihme commented on engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:
```diff
@@ -378,11 +378,7 @@ public void shouldCreateVariablesInProcessScope() {
-                    r ->
-                        r.getKey() == processInstanceKey
-                            && r.getIntent() == ProcessInstanceIntent.ELEMENT_COMPLETED)
+                .limitToProcessInstance(processInstanceKey)
```
:question: I don't understand how this works, to be honest. The `limitToProcessInstance` gives me all records where the key is the process instance (and before it is being closed). I would assume this predicate is positive for all process instance records. But somehow there seem to be variable records among them, or else this test doesn't work.
Are the variable records which have as key the process instance??? Because the key on a variable records points to a scope and the key on a process instance record is the key of the process instance?

Overall, I find this difficult to comprehend. This test is about the variables that are being written. Sp is there no way to rewrite this as:

`
RecordingExporter.variableRecords().withEventScopeKey(processInstanceKey).withIntent(VariableIntent.CREATED).findFirst()
`

saig0 replied:
>  I don't understand how this works, to be honest. The limitToProcessInstance gives me all records where the key is the process instance

No, your assumption is not correct. The method uses the `limit()` function of the stream. It returns all records (of all types) until the `completed` event of the process instance.

> is there no way to rewrite this as:

Sure, we could write the condition in this way. But this doesn't verify that any other variable not written, for example, for another scope.

pihme reviewed (changes requested):
Almost there. I still have some nitpicky comments/questions and one change request.

saig0 commented on engine/src/main/java/io/camunda/zeebe/engine/processing/ProcessEventProcessors.java:
```diff
@@ -201,10 +202,11 @@ private static void addProcessInstanceCreationStreamProcessors(
       final VariableBehavior variableBehavior) {
     final MutableElementInstanceState elementInstanceState = zeebeState.getElementInstanceState();
     final KeyGenerator keyGenerator = zeebeState.getKeyGenerator();
+    final ProcessEngineMetrics metrics = new ProcessEngineMetrics(zeebeState.getPartitionId());
```
I extracted the metrics and reuse them in both places.

saig0 commented:
@pihme I addressed your comments. Let's approve the PR.

pihme reviewed (approved):
bors merge

saig0 commented:
bors retry

pihme commented:
bors retry

saig0 commented:
The build failed repeatably with: 

```
Resolved Java 17.0.3+7 from tool-cache
Setting Java 17.0.3+7 as the default
Java configuration:
  Distribution: temurin
  Version: [17](https://github.com/camunda/zeebe/runs/6859178109?check_suite_focus=true#step:3:18).0.3+7
  Path: /opt/hostedtoolcache/Java_Temurin-Hotspot_jdk/17.0.3-7/x64
Creating settings.xml with server-id: github
Writing to /home/runner/.m2/settings.xml
Error: getCacheEntry failed: Cache service responded with 503
```

saig0 commented:
bors merge

pihme commented:
bors merge
